tnat64.c

```diff
@@ -415,54 +429,91 @@ int getsockname(GETSOCKNAME_SIGNATURE)
+
     show_msg(MSGDEBUG, "Got getsockname call for socket %d\n", __fd);
-    socklen_t needlen = *__len;
-    int ret = realgetsockname(__fd, __addr, &needlen);
-    if (ret < 0)
-    {
-        return ret;
-    }
 
-    if (*__len < sizeof(struct sockaddr_in))
-    {
-        *__len = sizeof(struct sockaddr_in);
-        errno = EINVAL;
-        return -1;
-    }
+    struct sockaddr_storage sockaddr_st;
+    socklen_t size = sizeof(sockaddr_st);
+    int ret = realgetsockname(__fd, (struct sockaddr *)&sockaddr_st, &size);
+    if (ret < 0) {
+        // If we end up here, it's not because of a too-small buffer, 
+        // because sockaddr_storage is the largest possible one.
+        return ret;
+    }   
 
     /* TODO: AF_INET6 is not necessarily 10, this debug print is wrong */
-    if (__addr->sa_family <= 10) {
-        show_msg(MSGDEBUG, "Address family is %s\n", afs[__addr->sa_family]);
+    if (sockaddr_st.ss_family <= 10) {
+        show_msg(MSGDEBUG, "Address family is %s\n", afs[sockaddr_st.ss_family]);
     }
 
-    if (__addr->sa_family == AF_INET6)
+    if (sockaddr_st.ss_family == AF_INET6)
     {
+
         struct sockaddr_in6 realsock;
         socklen_t realsocklen = sizeof(realsock);
-        int ret = realgetsockname(__fd, (struct sockaddr *)&realsock, &realsocklen);
-        if (ret < 0) {
-            return ret;
-        }
+        memcpy(&realsock, &sockaddr_st, realsocklen);
+
         if ((!memcmp(&realsock.sin6_addr, &ipv4mapped, NAT64PREFIXLEN)) || (check_prefix(config, &realsock.sin6_addr)))
         {
-            struct sockaddr_in * result;
-            result = (struct sockaddr_in *)__addr;
-            result->sin_family = AF_INET;
-            result->sin_port = realsock.sin6_port;
-            memcpy(&result->sin_addr, &realsock.sin6_addr.s6_addr[12], sizeof(struct in_addr));
+            struct sockaddr_in result;
+            memset(&result, 0, sizeof(result));
+
+            result.sin_family = AF_INET;
+            result.sin_port = realsock.sin6_port;
+            memcpy(&result.sin_addr, &realsock.sin6_addr.s6_addr[12], sizeof(struct in_addr));
+            
+            // Copy up to *__len bytes into the available space.
+            int memcpy_size = *__len;
+            if (memcpy_size > sizeof(struct sockaddr_in)) {
+                memcpy_size = sizeof(struct sockaddr_in);
+            }
+
+            memcpy(__addr, &result, memcpy_size);
+            *__len = sizeof(struct sockaddr_in);
+            return ret;
+        }
+        else {
+            // Application called getsockname, but the socket is not listening on an IPv6-mapped address. 
+            // The socket is listening on a "real" IPv6 address. 
+            // Returning that address to the application as-is is going to cause issues. 
+            // It's probably better to make the application believe it is bound to the unspecific address, 
+            // i.e. return an IPv4 address of 0.0.0.0 in this case. 
```

Comment:
Won’t this break applications talking to both IPv4 and IPv6?

Reply:
That's a very good point, that's probably true. I've also found another oddity in the code that could break applications that talk both IPv4 and IPv6. 

I guess the only way to 100% prevent these kinds of errors would be to keep a list of all sockets the application opened through the `socket` function, and take a note whether the socket originally was an IPv6 socket, or if it was an IPv4 socket that got "upgraded" to IPv6 through the library. Then it could, for "real" IPv6 sockets just return the original getsockname/getpeername, and for "fake" IPv6 sockets (where the application thinks it opened an IPv4 socket) it can fake the address 0.0.0.0.

Reply:
This issue should now be fixed with the updated patch, see my comment below.